setup: Use python3.10 (if available) for stbt_version 33

drothlis opened the pull request:
TODO:

- [x] Update `pip_deps` for v33 (see https://github.com/stb-tester/stb-tester/commit/7ac5997).
- [x] Test pytest 6.2.5 with VS Code.

drothlis committed:
setup: Use python3.10 (if available) for stbt_version 33

wmanley commented:
It would be good to have a `--reset` argument to `./stbt_rig.py setup` that would blow away `.venv` and recreate it for people upgrading from 32 to 33.

drothlis commented:
Is there any reason not to blow away `.venv` every time you run `./stbt_rig.py setup`? i.e. do you ever need to re-run `setup` without upgrading stbt_version?

drothlis committed:
PyPI package: description: Add copyright, links to manual

drothlis committed:
PyPI package: Update dependencies for stb-tester v33

drothlis committed:
PyPI package: Update version to 33.0.0

In preparation for stb-tester v33 release. We'll version stbt_rig
according to the latest stb-tester version that it supports.

drothlis committed:
setup: Update pip dependencies for stbt_version 33

Note that we don't need to specify pylint & its dependencies because
they are already specified by the stbt_core package (which pins them to
specific versions matching Ubuntu 22.04).

drothlis committed:
Silence some pylint warnings seen on my dev machine

CI doesn't warn about these because we're using an older version of
pylint there, to support Python 2.

drothlis committed:
Remove unused CI Dockerfiles

Added in commit 26844b0: "Add dockerfiles to run tests locally
with the same dependencies installed as Travis". We don't use these
anymore as we have learned how to use virtualenvs. ;-)

drothlis committed:
setup: Re-create venv with Python 3.10 when upgrading to stbt_version 33

If the venv's Python version isn't 3.10, *and* Python 3.10 is actually
available, we delete the ".venv" directory and re-create it with Python
3.10.

drothlis committed:
setup: Refactor 1st & 2nd stages into separate functions

Just so I get more horizontal space, really.

drothlis commented:
On Ubuntu 20.04, which has Python 3.8, I had installed Python 3.10 from the deadsnakes PPA but I got this error running `stbt_rig.py setup`:

```
❯ ./stbt_rig.py setup --vscode
stbt_rig.py: WARNING: Virtualenv in '/home/drothlis/work/stb-tester.com/test-packs/demo/.venv' has Python version 3.6.15; deleting .venv and recreating it with Python 3.10.
Error: Command '['/home/drothlis/work/stb-tester.com/test-packs/demo/.venv/bin/python3.10', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1.
Traceback (most recent call last):
  File "./stbt_rig.py", line 1925, in <module>
    sys.exit(main())
  File "./stbt_rig.py", line 83, in main
    return main_with_args(args)
  File "./stbt_rig.py", line 110, in main_with_args
    return cmd_setup(args, args.node_id)
  File "./stbt_rig.py", line 544, in cmd_setup
    setup_stage1(this_stbt_rig, root)
  File "./stbt_rig.py", line 584, in setup_stage1
    cwd=root)
  File "/usr/lib/python2.7/subprocess.py", line 190, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '[u'python3.10', u'-m', u'venv', u'.venv']' returned non-zero exit status 1
```

The solution was to install `python3.10-venv` (see https://stackoverflow.com/a/71726397).

drothlis commented:
Then it fails because we haven't published stb-tester 33 yet:

> ERROR: Could not find a version that satisfies the requirement stb-tester~=33.0

@wmanley please review, I'd like to merge it anyway. This only kicks in if you set `stbt_version = 33` in your `.stbt.conf`.

drothlis commented:
Testing with `stbt_version = 32`:

```
❯ ./stbt_rig.py setup --vscode
stbt_rig.py: WARNING: Virtualenv in '/home/drothlis/work/stb-tester.com/test-packs/demo/.venv' has Python version 3.10.5; deleting .venv and recreating it with Python 3.6.
Collecting pip
  Using cached https://files.pythonhosted.org/packages/a4/6d/6463d49a933f547439d6b5b98b46af8742cc03ae83543e4d7688c2420f8b/pip-21.3.1-py3-none-any.whl
Installing collected packages: pip
  Found existing installation: pip 18.1
    Uninstalling pip-18.1:
      Successfully uninstalled pip-18.1
Successfully installed pip-21.3.1
[...]
Successfully installed SecretStorage-3.3.2 astroid-1.6.0 atomicwrites-1.4.0 attrs-20.2.0 backports.zoneinfo-0.2.1 certifi-2022.6.15 cffi-1.15.0 charset-normalizer-2.0.12 cryptography-37.0.2 future-0.15.2 idna-3.3 importlib-metadata-4.8.3 importlib-resources-5.4.0 isort-4.3.4 jeepney-0.7.1 keyring-23.4.1 lazy-object-proxy-1.7.1 mccabe-0.7.0 more-itertools-8.13.0 numpy-1.19.5 opencv-python-3.4.18.65 packaging-21.3 pluggy-0.13.1 py-1.11.0 pycparser-2.21 pylint-1.8.3 pyparsing-3.0.9 pytest-4.6.11 pytz-deprecation-shim-0.1.0.post0 requests-2.27.1 six-1.16.0 stb-tester-32.2.0 stbt-core-32.0.4 typing-extensions-4.1.1 tzdata-2022.1 tzlocal-4.2 urllib3-1.26.9 wcwidth-0.2.5 wrapt-1.14.1 zipp-3.6.0
[...]
```

Running it again:

```
❯ ./stbt_rig.py setup --vscode
Requirement already satisfied: pip in ./.venv/lib/python3.6/site-packages (21.3.1)
[...]
```

drothlis commented:
### Test pytest 6.2.5 with VS Code

Tested by temporarily changing the v32 `pip_deps`, re-running `./stbt_rig.py setup --vscode`, and ran a test on demo.stb-tester.com from VSCode. The "Python Test Log" output showed:

> platform linux -- Python 3.6.15, pytest-6.2.5, py-1.11.0, pluggy-0.13.1

I also tested stbt_rig's snapshots by modifying the test before running it (though I think snapshots don't involve pytest anyway).

wmanley commented on stbt_rig.py:
```diff
@@ -554,40 +554,34 @@ def cmd_setup(args, node_id):
+                    "Virtualenv in '%s' has Python version %s; "
+                    "deleting .venv and recreating it with Python %s.",
+                    venv_dir, venv_python_version, v)
+                shutil.rmtree(venv_dir)
```
I think it would be good to prompt before running this.  The user might want to back up the `.venv` or record what versions of things they have installed there.

Untested:

```suggestion
                while True:
                    ok = input("Delete and recreate .venv?: Y/N: ").upper().strip()
                    if ok == "Y":
                        shutil.rmtree(venv_dir)
                        break
                    elif ok = "N":
                        break
```

drothlis replied:
Done in c49beb8.

Tested by setting stbt_version to 33, running `./stbt_rig.py setup`, answering `n` and then running it again and answering `y`.

drothlis committed:
setup: Prompt before deleting .venv

drothlis commented:
More testing: Temporarily edited stbt_rig.py to require "3.100" instead of "3.10", and ran `./stbt_rig.py setup` to ensure it warns but doesn't try to delete `.venv`:

```
❯ ./stbt_rig.py setup 
stbt_rig.py: WARNING: Using Python version 3.8.10 which may not be fully compatible with stb-tester 33. For best compatibility install Python 3.100.
Requirement already satisfied: pip in ./.venv/lib/python3.6/site-packages (21.3.1)
[...]
```
